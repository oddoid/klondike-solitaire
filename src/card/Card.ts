import {Rank} from './Rank'
import {Suit} from './Suit'

/** A playing card. */
export type Card = {
  readonly suit: Suit
  readonly rank: Rank
  direction: Card.Direction
}

export namespace Card {
  /**
   * [Playing Cards Unicode range](https://en.wikipedia.org/wiki/Standard_52-card_deck#Unicode).
   */
  const unicode = <const>{rangeStart: 0x1f0a0, rankSize: 16, suitMax: 3}

  /**
   * The physical orientation of the card. All card backs are identical and
   * occlude the fronts. The front, or face, shows the card suit and rank. Each
   * card has a unique suit-rank identity that is secret when hidden and known
   * when visible.
   */
  export type Direction = 'down' | 'up'

  /** Unicode to Card adapter. Only face-up codes are supported. */
  export function fromString(code: string, direction: Direction = 'up'): Card {
    const point = (code.codePointAt(0) ?? 0) - unicode.rangeStart
    const suit = Suit.fromOrder[unicode.suitMax - ~~(point / unicode.rankSize)]
    const rank = Rank.fromPoint[point % unicode.rankSize]
    if (suit === undefined || rank === undefined)
      throw new Error(`Unknown point for "${code}".`)
    return {suit, rank, direction}
  }

  /**
   * @arg directed When true, respect card direction. That is, face down cards
   *  are represented as card backs. When false, all card identities are
   *  visible regardless of card directions. For example, a face down ace of
   *  spades is `'ðŸ‚ '` when `directed` is true and `'ðŸ‚¡'` when false.
   * @ret The Unicode character representation for each card.
   */
  export function toString(
    directed: boolean,
    ...cards: readonly Readonly<Card>[]
  ): string {
    return cards.reduce((str, card) => str + cardToString(directed, card), '')
  }

  function cardToString(directed: boolean, card: Readonly<Card>): string {
    if (directed && card.direction === 'down') return 'ðŸ‚ '
    const point =
      unicode.rangeStart +
      (unicode.suitMax - Suit.toOrder[card.suit]) * unicode.rankSize +
      Rank.toPoint[card.rank]
    return String.fromCodePoint(point)
  }
}
