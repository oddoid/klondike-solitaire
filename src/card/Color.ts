import { Immutable } from '@/oidlib';

/** Suit primary hue. */
export type Color = Parameters<typeof Color.values['has']>[0];

export namespace Color {
  export const values = Immutable(new Set(['Black', 'Red'] as const));
}
