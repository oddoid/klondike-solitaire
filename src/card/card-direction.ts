/**
 * The physical orientation of the card. All card backs are identical and
 * occlude the fronts. The front, or face, shows the card suit and rank. Each
 * card has a unique suit-rank identity that is secret when hidden (down) and
 * known when visible (up).
 */
export type CardDirection = Parameters<typeof CardDirectionSet['has']>[0]

export const CardDirectionSet = new Set(['Down', 'Up'] as const)
