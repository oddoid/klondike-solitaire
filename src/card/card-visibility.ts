/**
 * When directed, respect card direction. That is, face down cards are
 * represented as card backs. When false, all card identities are visible
 * regardless of card directions. For example, a face down ace of spades is
 * `'ðŸ‚ '` when `'directed'` and `'ðŸ‚¡'` when `'undirected'`.
 */
export type CardVisibility = Parameters<(typeof CardVisibilitySet)['has']>[0]
export const CardVisibilitySet = new Set(<const>['Directed', 'Undirected'])
