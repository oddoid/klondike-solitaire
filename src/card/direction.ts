import { Immutable } from '@/oidlib'

/**
 * The physical orientation of the card. All card backs are identical and
 * occlude the fronts. The front, or face, shows the card suit and rank. Each
 * card has a unique suit-rank identity that is secret when hidden (down) and
 * known when visible (up).
 */
export type Direction = Parameters<typeof Direction.values['has']>[0]

export namespace Direction {
  export const values = Immutable(new Set(['Down', 'Up'] as const))
}
