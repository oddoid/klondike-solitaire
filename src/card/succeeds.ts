import { Card } from '@/solitaire';

/**
 * Returns true if right card directly succeeds left.
 *
 * Empty behavior (left or right is undefined) is implementation-specific but
 * the function is guaranteed to be called once when both inputs are empty and
 * otherwise once for every pair (len - 1).
 */
export type Succeeds = (
  lhs: Readonly<Card> | undefined,
  rhs: Readonly<Card> | undefined,
) => boolean;
