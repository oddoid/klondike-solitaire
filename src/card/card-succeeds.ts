import {Card} from './card.js'

/**
 * Returns true if right card directly succeeds left.
 *
 * Empty behavior (left or right is undefined) is implementation-specific but
 * the function is guaranteed to be called once when both inputs are empty and
 * otherwise once for every pair (len - 1).
 */
export type CardSucceeds = (
  lhs: Readonly<Card> | undefined,
  rhs: Readonly<Card> | undefined
) => boolean
