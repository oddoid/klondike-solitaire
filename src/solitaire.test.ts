import { assertAlmostEquals, assertEquals } from 'std/testing/asserts.ts'
import { assertSnapshot } from 'std/testing/snapshot.ts'
import { cardToString } from './card/card.ts'
import { foundationToString } from './layout/foundation.ts'
import {
  shuffle,
  Solitaire,
  solitaireBuild,
  solitaireDeal,
  solitairePoint,
  solitaireToString,
  swapIndices,
  uncapitalize,
} from './solitaire.ts'

Deno.test('Deal', async (test) => {
  let solitaire: Solitaire

  await test.step(
    'Set the game.',
    async (test) => {
      solitaire = Solitaire(() => 1 - Number.EPSILON)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step('The foundation piles are empty.', (test) => {
    assertEquals(cardToString('Directed', ...solitaire.foundation[0]), '')
    assertEquals(cardToString('Directed', ...solitaire.foundation[1]), '')
    assertEquals(cardToString('Directed', ...solitaire.foundation[2]), '')
    assertEquals(cardToString('Directed', ...solitaire.foundation[3]), '')
    assertSnapshot(test, foundationToString(solitaire.foundation, 'Directed'), {
      dir: '.',
    })
  })

  await test.step('The waste is empty.', () => {
    assertEquals(solitaire.waste.length, 0)
  })

  await test.step(
    'Deal.',
    async (test) => {
      solitaireDeal(solitaire)
      await assertSnapshot(test, solitaireToString(solitaire, 'Undirected'), {
        dir: '.',
      })
    },
  )

  await test.step('The stock is dealt into the tableau.', () => {
    assertEquals(
      cardToString('Directed', ...solitaire.stock),
      // 🃑🃒🃓🃔🃕🃖🃗🃘🃙🃚🃛🃝🃞   🃁🃂🃃🃄🃅🃆🃇🃈🃉🃊🃋
      '🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠' + '🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠',
    )
    assertEquals(
      solitaire.tableau.map((pile) => cardToString('Directed', ...pile)),
      [
        '🂠', //       🂮
        '🂠🂠', //      🂫🂭
        '🂠🂠🂠', //     🂨🂩🂪
        '🂠🂠🂠🂠', //    🂤🂥🂦🂧
        '🂠🂠🂠🂠🂠', //   🂽🂾🂡🂢🂣
        '🂠🂠🂠🂠🂠🂠', //  🂶🂷🂸🂹🂺🂻
        '🂠🂠🂠🂠🂠🂠🂠', // 🃍🃎🂱🂲🂳🂴🂵
      ],
    )
  })

  await test.step(
    'Draw all.',
    async (test) => {
      while (solitaire.stock.length > 0) {
        solitairePoint(solitaire, solitaire.stock.at(-1)!)
        await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
      }
    },
  )

  await test.step(
    'Redeal.',
    async (test) => {
      solitaireDeal(solitaire)
      await assertSnapshot(test, solitaireToString(solitaire, 'Undirected'), {
        dir: '.',
      })
    },
  )

  await test.step(
    'Draw all.',
    async (test) => {
      while (solitaire.stock.length > 0) {
        solitairePoint(solitaire, solitaire.stock.at(-1)!)
        await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
      }
    },
  )
})

Deno.test('Playthrough', async (test) => {
  let solitaire!: Solitaire
  await test.step(
    'Set the game.',
    async (test) => {
      solitaire = Solitaire(() => 1 - Number.EPSILON, 0, 3, 7)
      solitaireDeal(solitaire)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  for (const [index, column] of solitaire.tableau.entries()) {
    await test.step(
      `Reveal tableau pile ${index}.`,
      async (test) => {
        solitairePoint(solitaire, column.at(-1)!)
        await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
      },
    )
  }

  await test.step(
    'Take the ten of spades from tableau 2, top.',
    async (test) => {
      solitairePoint(solitaire, solitaire.tableau[2]!.at(-1)!)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Put onto the jack of hearts at tableau 5.',
    async (test) => {
      solitaireBuild(solitaire, { type: 'Tableau', x: 5 })
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Take the the jack of hearts at tableau 5, top.',
    async (test) => {
      solitairePoint(solitaire, solitaire.tableau[5]!.at(-2)!)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Put onto the queen of spades at tableau 1.',
    async (test) => {
      solitaireBuild(solitaire, { type: 'Tableau', x: 1 })
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Reveal tableau pile 2.',
    async (test) => {
      solitairePoint(solitaire, solitaire.tableau[2]!.at(-1)!)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Reveal tableau pile 5.',
    async (test) => {
      solitairePoint(solitaire, solitaire.tableau[5]!.at(-1)!)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Take the the nine of spades at tableau 2, top.',
    async (test) => {
      solitairePoint(solitaire, solitaire.tableau[2]!.at(-1)!)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Put onto the ten of hearts at tableau 1.',
    async (test) => {
      solitaireBuild(solitaire, { type: 'Tableau', x: 5 })
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )

  await test.step(
    'Reveal tableau pile 2.',
    async (test) => {
      solitairePoint(solitaire, solitaire.tableau[2]!.at(-1)!)
      await assertSnapshot(test, solitaireToString(solitaire), { dir: '.' })
    },
  )
})

Deno.test('Shuffle: permutations.', () => {
  const iterations = 1_000_000
  const array = ['a', 'b', 'c', 'd', 'e']
  const permutations = permute(array).map((permutation) => permutation.join(''))

  const distribution: Record<string, number> = {}
  for (let i = 0; i < iterations; i++) {
    shuffle(array, Math.random)
    const permutation = array.join('')
    distribution[permutation] ??= 0
    distribution[permutation]++
  }

  for (const permutation of permutations) {
    const occurrences = distribution[permutation] ?? 0
    const frequency = occurrences / iterations
    assertAlmostEquals(frequency, 1 / permutations.length, 0.001)
  }
})

Deno.test('Shuffle: no randomization.', () => {
  const alphabet = 'abcdefghijklmnopqrstuvwxyz'
  const letters = [...alphabet]
  shuffle(letters, () => 1 - Number.EPSILON)
  assertEquals(letters.join(''), alphabet)
})

/** [Heap's_algorithm](https://en.wikipedia.org/wiki/Heap%27s_algorithm). */
function permute<T>(array: T[], n: number = array.length): T[][] {
  if (n === 1) return [[...array]]

  const permutations = []
  permutations.push(...permute(array, n - 1))

  for (let i = 0; i < n - 1; i++) {
    if (n % 2) swapIndices(array, 0, n - 1)
    else swapIndices(array, i, n - 1)
    permutations.push(...permute(array, n - 1))
  }

  return permutations
}

Deno.test('Str', async (test) => {
  for (
    const [str, uncapitalized] of [
      ['', ''],
      [' ', ' '],
      ['\t', '\t'],
      ['\n', '\n'],
      ['   ', '   '],
      ['a', 'a'],
      ['A', 'a'],
      ['abc', 'abc'],
      ['ABC', 'aBC'],
    ] as const
  ) {
    await test.step(`strUncapitalize(${str}) => ${uncapitalized}`, () =>
      assertEquals(uncapitalize(str), uncapitalized))
  }
})
