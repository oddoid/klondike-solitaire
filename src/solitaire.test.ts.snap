export const snapshot = {};

snapshot[`Deal > Set the game. 1`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Deal > The foundation piles are empty. 1`] = `"🃟🃟🃟🃟"`;

snapshot[`Deal > Deal. 1`] = `
"🃟🃟🃟🃟
🂮🂫🂨🂤🂽🂶🃍
　🂭🂩🂥🂾🂷🃎
　　🂪🂦🂡🂸🂱
　　　🂧🂢🂹🂲
　　　　🂣🂺🂳
　　　　　🂻🂴
　　　　　　🂵
🃑🃒🃓🃔🃕🃖🃗🃘🃙🃚🃛🃝🃞🃁🃂🃃🃄🃅🃆🃇🃈🃉🃊🃋 🃟 🃟"
`;

snapshot[`Deal > Draw all. 1`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃋🃊🃉 🃟"
`;

snapshot[`Deal > Draw all. 2`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃈🃇🃆 🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 3`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃅🃄🃃 🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 4`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃂🃁🃞 🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 5`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃝🃛🃚 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 6`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠 🃙🃘🃗 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 7`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠 🃖🃕🃔 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 8`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
 🃓🃒🃑 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Redeal. 1`] = `
"🃟🃟🃟🃟
🂮🂫🂨🂤🂽🂶🃍
　🂭🂩🂥🂾🂷🃎
　　🂪🂦🂡🂸🂱
　　　🂧🂢🂹🂲
　　　　🂣🂺🂳
　　　　　🂻🂴
　　　　　　🂵
🃑🃒🃓🃔🃕🃖🃗🃘🃙🃚🃛🃝🃞🃁🃂🃃🃄🃅🃆🃇🃈🃉🃊🃋 🃟 🃟"
`;

snapshot[`Deal > Draw all. 9`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃋🃊🃉 🃟"
`;

snapshot[`Deal > Draw all. 10`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃈🃇🃆 🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 11`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃅🃄🃃 🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 12`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃂🃁🃞 🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 13`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃝🃛🃚 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 14`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠 🃙🃘🃗 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 15`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠 🃖🃕🃔 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Deal > Draw all. 16`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
 🃓🃒🃑 🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠"
`;

snapshot[`Playthrough > Set the game. 1`] = `
"🃟🃟🃟🃟
🂠🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 0. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂠🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 1. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 2. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　🂪🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 3. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　🂪🂠🂠🂠🂠
　　　🂧🂠🂠🂠
　　　　🂠🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 4. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　🂪🂠🂠🂠🂠
　　　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　🂠🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 5. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　🂪🂠🂠🂠🂠
　　　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　🂻🂠
　　　　　　🂠
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 6. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　🂪🂠🂠🂠🂠
　　　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　🂻🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Take the ten of spades from tableau 2, top. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　🂻🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟
🂪 from Tableau (2, 2)"
`;

snapshot[`Playthrough > Put onto the jack of hearts at tableau 5. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　🂻🂠
　　　　　🂪🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Take the the jack of hearts at tableau 5, top. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　　　🂠🂠🂠🂠
　　　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　　🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟
🂻🂪 from Tableau (5, 5)"
`;

snapshot[`Playthrough > Put onto the queen of spades at tableau 1. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂠🂠🂠🂠🂠
　🂻　🂠🂠🂠🂠
　🂪　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　　🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 2. 2`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂩🂠🂠🂠🂠
　🂻　🂠🂠🂠🂠
　🂪　🂧🂠🂠🂠
　　　　🂣🂠🂠
　　　　　　🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 5. 2`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭🂩🂠🂠🂠🂠
　🂻　🂠🂠🂠🂠
　🂪　🂧🂠🂠🂠
　　　　🂣🂺🂠
　　　　　　🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Take the the nine of spades at tableau 2, top. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭　🂠🂠🂠🂠
　🂻　🂠🂠🂠🂠
　🂪　🂧🂠🂠🂠
　　　　🂣🂺🂠
　　　　　　🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟
🂩 from Tableau (2, 1)"
`;

snapshot[`Playthrough > Put onto the ten of hearts at tableau 1. 1`] = `
"🃟🃟🃟🃟
🂮🂠🂠🂠🂠🂠🂠
　🂭　🂠🂠🂠🂠
　🂻　🂠🂠🂠🂠
　🂪　🂧🂠🂠🂠
　　　　🂣🂺🂠
　　　　　🂩🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;

snapshot[`Playthrough > Reveal tableau pile 2. 3`] = `
"🃟🃟🃟🃟
🂮🂠🂨🂠🂠🂠🂠
　🂭　🂠🂠🂠🂠
　🂻　🂠🂠🂠🂠
　🂪　🂧🂠🂠🂠
　　　　🂣🂺🂠
　　　　　🂩🂠
　　　　　　🂵
🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠🂠 🃟 🃟"
`;
